addons/sale_coupon/tests/test_program_numbers.py

```diff
@@ -1082,3 +1082,90 @@ def test_program_free_prods_with_min_qty_and_reward_qty_and_rule(self):
 
         self.assertEqual(order.amount_total, 500.0, 'The price must be 500.0 since two conference chairs are free and the user only bought one')
         self.assertEqual(order.order_line[2].price_total, -100.0, 'The last order line should apply a reduction of 100.0 since there is one conference chair that cost 100.0')
+
+    def test_program_free_product_different_than_rule_product_with_multiple_application(self):
+        order = self.empty_order
+        product_D = self.env['product.product'].create({
+            'name': 'Product D',
+            'list_price': 100,
+            'sale_ok': True,
+            'taxes_id': [(6, 0, [])],
+
+        })
+        product_E = self.env['product.product'].create({
+            'name': 'Product E',
+            'list_price': 5,
+            'sale_ok': True,
+            'taxes_id': [(6, 0, [])],
+
+        })
+        immediate_program = self.env['sale.coupon.program'].create({
+            'name': '3 Product D = 1 free Product E',
+            'promo_code_usage': 'no_code_needed',
+            'reward_type': 'product',
+            'reward_product_id': product_E.id,
+            'rule_products_domain': "[('id', 'in', [%s])]" % (product_D.id),
+            'active': True,
+        })
+
+        sol_D = self.env['sale.order.line'].create({
+            'product_id': product_D.id,
+            'product_uom_qty': 6.0,
+            'order_id': order.id,
+        })
+        sol_E = self.env['sale.order.line'].create({
+            'product_id': product_E.id,
+            'product_uom_qty': 1.0,
+            'order_id': order.id,
+        })
+
+        order.recompute_coupon_lines()
+
+        self.assertEqual(len(order.order_line), 3, 'The order must contain 3 order lines including one for a free Product E')
+        self.assertEqual(order.amount_total, 600.0, 'The price must be 600.0 since the Product E is free')
+        self.assertEqual(order.order_line[2].product_uom_qty, 1, 'The last order line should only have a quantity of 1 since there is only 1 Product E')
+
+        sol_E.product_uom_qty = 2
+
+        order.recompute_coupon_lines()
+
+        self.assertEqual(len(order.order_line), 3, 'The order must contain 3 order lines including one for a free Product E')
+        self.assertEqual(order.amount_total, 600.0, 'The price must be 600.0 since the 2 Product E are free')
+        self.assertEqual(order.order_line[2].product_uom_qty, 2, 'The last order line should have a quantity of 2 since there 2 Product E and 6 Product D')
+
+    def test_program_modify_reward_line_qty(self):
+        order = self.empty_order
+        product_F = self.env['product.product'].create({
+            'name': 'Product F',
+            'list_price': 100,
+            'sale_ok': True,
+            'taxes_id': [(6, 0, [])],
+
+        })
+        immediate_program = self.env['sale.coupon.program'].create({
```

Comment:
unused variable, check to configure your linter

Reply:
I know this isn't good, but I saw it was done this way throughout a lot of tests. I didn't know if it was better to "keep the existing style", or just follow the linter warnings.

Reply:
I will let you blame the file and see when the linter error where introduced ;-)
Those coupon tests files were perfectly linted for years.

People just don't have a linter those days or just don't care :shrug: 

But yeah, your priority should be to avoid linter error.
Some people's IDE will surround the lines of code that have linting error, some will background-color it. Some people will have a warning on file save about the linting error (even the existing one). Some team even added an auto-linter on save for the whole file.
The point is, avoid writing new ones, it's very bad ;-)

(Couldn't send the message the other day, as github was down)

Reply:
Alright, I'll keep that in mind ! Thank you for the info ! :+1: 
Does everything looks good now ?